src/include/txn_inline.h

```diff
@@ -901,10 +904,14 @@ __wt_txn_read_upd_list_internal(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt,
          * Save the prepared update to help us detect if we race with prepared commit or rollback
          * irrespective of update visibility.
          */
-        if ((prepare_state == WT_PREPARE_INPROGRESS || prepare_state == WT_PREPARE_LOCKED) &&
+        if ((prepare_state == WT_PREPARE_INPROGRESS ||
+              prepare_state == WT_PREPARE_ROLLBACK_INPROGRESS ||
+              prepare_state == WT_PREPARE_LOCKED) &&
           prepare_updp != NULL && *prepare_updp == NULL &&
-          F_ISSET(upd, WT_UPDATE_PREPARE_RESTORED_FROM_DS))
+          F_ISSET(upd, WT_UPDATE_PREPARE_RESTORED_FROM_DS)) {
             *prepare_updp = upd;
+            *prepare_statep = prepare_state;
```

Comment:
We are already returning the prepared update back to the caller, do we need a separate prepared state flag also?

Reply:
It is repetitive. We have asserted before.

Reply:
We need to record its prepared state during the walk of the update chain and compare later to see if it changes.